detect.go

```diff
@@ -10,23 +10,60 @@ import (
 )
 
 //go:generate faux --interface Parser --output fakes/parser.go
+
+// Parser defines the interface for determining if the Gemfile contains the
+// "rails" gem.
 type Parser interface {
 	Parse(path string) (hasRails bool, err error)
 }
 
+// BuildPlanMetadata declares the set of metadata included in build plan
+// requirements.
 type BuildPlanMetadata struct {
+
+	// Build is set to true when the build plan requirement should be made
+	// available during the build phase of the buildpack lifecycle.
 	Build bool `toml:"build"`
 }
 
+// Detect will return a packit.DetectFunc that will be invoked during the
+// detect phase of the buildpack lifecycle.
+//
+// The detection criteria is twofold:
+//   1. An assets directory must be present in the application source code.
+//   These directories include app/assets, lib/assets, vendor/assets, and
+//   app/javascript.
+//   2. The Gemfile must reference the "rails" gem.
+//
+// If both of these criteria are met, then the buildpack will require "node",
+// "mri", "bundler", and "gems" as build-time build plan requirements.
+//
+// Additionally, for Rails 6, we want to run yarn install ahead of asset
+// compilation. We can detect this case by the presence of a yarn.lock file. In
+// that case, the buildpack will also require "node_modules" as a build-time
+// build plan requirement.
 func Detect(gemfileParser Parser) packit.DetectFunc {
 	return func(context packit.DetectContext) (packit.DetectResult, error) {
-		_, err := os.Stat(filepath.Join(context.WorkingDir, "app", "assets"))
-		if err != nil {
-			if errors.Is(err, os.ErrNotExist) {
-				return packit.DetectResult{}, packit.Fail
+		hasAssetsDirectory := false
+		for _, path := range []string{
+			filepath.Join(context.WorkingDir, "app", "assets"),
+			filepath.Join(context.WorkingDir, "lib", "assets"),
+			filepath.Join(context.WorkingDir, "vendor", "assets"),
+			filepath.Join(context.WorkingDir, "app", "javascript"),
+		} {
+			_, err := os.Stat(path)
+			if err == nil {
+				hasAssetsDirectory = true
+				break
+			} else {
+				if !errors.Is(err, os.ErrNotExist) {
+					return packit.DetectResult{}, fmt.Errorf("failed to stat app/assets: %w", err)
```

Comment:
This should refer to the variable `path` instead of `app/assets`

Reply:
I didn't understand. This part of code tries to find assets in any of four folders: app/assets, app/javascript, lib/assets, or vendor/assets. If any of these folders exist, rails-assets DETECT phase is successful and CNB will participate in BUILD phase. This is how it works in pakito's CNB since last Friday. If we want to change the logic (I can't be a judge here because of my tiny knowledge of Rails), please tell me which way. But if this logic is OK, I prefer to maintain code base as close to pakito as it possible. I'm sure they'll change it next months many ways. Last updates regarding list of folders with assets didn't break anything, but extended it from 1 to 4 folders.

Reply:
@ybykov-a9s My comment was only about the error message. The error message produced by `fmt.Errorf("failed to stat app/assets: %w", err)` is misleading because it could any of the paths specified in variable `path`. So IMO this error message should use the variable `path`. Do you see what I mean?

Reply:
I got it. Again, copied from original buildpack but missed texts. Fixed.